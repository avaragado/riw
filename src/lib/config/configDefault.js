// @flow

const cfg: RIWConfig = {
    configFile: null, // overridden
    rootDir: '', // overridden
    defaultLocale: 'en-us',
    targetLocales: [],
    translationsDatabaseFile: 'src/locale/riw-db.json',
    sourceDirs: ['src/**/*.js'],
    collateDir: 'tmp/babel-plugin-react-intl',
    inputMode: 'source',
    translationsOutputFile: 'src/locale/[locale].json',
    outputMode: 'file-per-locale',
    todoFile: 'src/locale/TODO-untranslated.json',
};

export default cfg;
