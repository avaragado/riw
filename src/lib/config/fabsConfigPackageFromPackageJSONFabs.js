// @flow

import path from 'path';

export default (fabsPackageJSON: ?AbsolutePath): ?AbsolutePath => (
    fabsPackageJSON
        ? path.resolve(path.dirname(fabsPackageJSON), '.riw-config.js')
        : null
    );
