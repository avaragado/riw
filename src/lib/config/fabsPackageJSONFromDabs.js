// @flow

import findUp from 'find-up';

export default (cwd: AbsolutePath): ?AbsolutePath => findUp.sync('package.json', { cwd });

