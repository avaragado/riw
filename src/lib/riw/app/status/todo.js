// @flow

import path from 'path';

import compose from 'ramda/src/compose';
import groupBy from 'ramda/src/groupBy';
import prop from 'ramda/src/prop';

import type { LocaleId, UntranslatedMessageDescriptor } from '../../../../types';
import type { Config } from '../../../config';

import loadJSONOrNull from './loadJSONOrNull';

type Result = (config: Config) => {
    todo: null | {
        [key: LocaleId]: UntranslatedMessageDescriptor[],
    },
};

const result: Result = compose(
    todo => ({ todo }),
    (armdu: ?UntranslatedMessageDescriptor[]) => (armdu == null
        ? null
        : groupBy(prop('locale'), armdu)
    ),
    loadJSONOrNull,
    (config: Config) => path.resolve(
        config.rootDir,
        config.todoFile,
    ),
);

export default result;
