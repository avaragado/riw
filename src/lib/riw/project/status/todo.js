// @flow

import path from 'path';

import compose from 'ramda/src/compose';
import groupBy from 'ramda/src/groupBy';
import prop from 'ramda/src/prop';

import loadJSONOrNull from './loadJSONOrNull';

type Result = (config: RIWConfig) => {
    todo: null | {
        [key: LocaleId]: RIWMessageDescriptorUntranslated[],
    },
};

const result: Result = compose(
    todo => ({ todo }),
    (armdu: ?RIWMessageDescriptorUntranslated[]) => (armdu == null
        ? null
        : groupBy(prop('locale'), armdu)
    ),
    loadJSONOrNull,
    (config: RIWConfig) => path.resolve(
        config.rootDir || '.',
        config.todoFile,
    ),
);

export default result;
