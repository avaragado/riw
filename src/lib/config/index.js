// @flow

import configDefault from './resolve/configDefault';
import fabsConfigPackageForCWD from './fromPackage/fabsConfigPackageForCWD';
import configFromPath from './fromPath';
import configFromPackage from './fromPackage';
import configResolve from './resolve';

export {
    configDefault,
    fabsConfigPackageForCWD,
    configFromPath,
    configFromPackage,
    configResolve,
};
