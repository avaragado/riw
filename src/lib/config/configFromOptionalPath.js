// @flow

import path from 'path';

import log from '../log';

import configDefault from './configDefault';
import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

const requireOrNull = (fabs: AbsolutePath): ?Object => {
    try {
        return require(fabs);

    } catch (err) {
        return null;
    }
};

const configFromPath = (pathConfig: Path): ?RIWConfig => {
    const fabs: AbsolutePath = path.resolve(pathConfig);
    const configOverride = requireOrNull(fabs);

    if (configOverride === null) {
        log.error('riw', 'File not found: %j.', fabs);
        return null;
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabs),
    };
};

const configFromPackage = (): ?RIWConfig => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsConfigPackage || !fabsPackageJSON) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    let configOverride: ?Object = requireOrNull(fabsConfigPackage);

    if (configOverride === null) {
        configOverride = requireOrNull(fabsPackageJSON);

        if (configOverride && configOverride.riw) {
            configOverride = configOverride.riw;
        } else {
            configOverride = null;
        }
    }

    if (configOverride === null) {
        log.warn('riw', 'No config located for the project. You probably want to add some configuration.');
        configOverride = {};
    }

    return {
        ...configDefault,
        ...configOverride,
        rootDir: path.dirname(fabsConfigPackage),
    };

};

export default (pathConfig: ?Path): ?RIWConfig => (
    pathConfig
        ? configFromPath(pathConfig)
        : configFromPackage()
);
