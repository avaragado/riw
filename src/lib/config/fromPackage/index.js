// @flow

import log from '../../log';

import fabsPackageJSONFromDabs from './fabsPackageJSONFromDabs';
import fabsConfigPackageFromPackageJSONFabs from './fabsConfigPackageFromPackageJSONFabs';

export default (): ?RIWConfigSparseWithSource => {
    const fabsPackageJSON = fabsPackageJSONFromDabs(process.cwd());
    const fabsConfigPackage = fabsConfigPackageFromPackageJSONFabs(fabsPackageJSON);

    if (!fabsPackageJSON || !fabsConfigPackage) {
        log.error('riw', 'Unable to find package root');
        return null;
    }

    try {
        return {
            ...(require(fabsConfigPackage): RIWConfigSparse),
            configFile: fabsConfigPackage,
        };

    } catch (err) {
        // fabsPackageJSON exists as we hunted the fs for it above:
        // but it might not have the riw key.
        const config: RIWConfigSparse = require(fabsPackageJSON).riw;

        if (!config) {
            log.warn('riw', 'No app config found. You probably want to add some configuration.');
        }

        return {
            ...config,
            configFile: fabsPackageJSON,
        };
    }
};
