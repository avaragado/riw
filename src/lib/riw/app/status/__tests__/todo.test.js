// @flow

import mock from 'mock-fs';

import { configResolve } from '../../../../config';

import todo from '../todo';

const stringify = obj => JSON.stringify(obj, null, 4);

const cfgBase: RIWConfigSparseWithSource = {
    todoFile: 'fixtures/dir/TODO.json',
};

type Fixture = {
    name: string,
    in: ?RIWMessageDescriptorUntranslated[],
};

const fixtures: Fixture[] = [
    {
        name: '01',
        in: null,
    },

    {
        name: '02',
        in: [
            { id: 'md1', defaultMessage: 'aa', description: 'aa desc', locale: 'xx-xx' },
            { id: 'md1', defaultMessage: 'aa', description: 'aa desc', locale: 'yy-yy' },
            { id: 'md1', defaultMessage: 'aa', description: 'aa desc', locale: 'zz-zz' },
            { id: 'md2', defaultMessage: 'bb', description: 'bb desc', locale: 'xx-xx' },
            { id: 'md3', defaultMessage: 'cc', description: 'cc desc', locale: 'zz-zz' },
            { id: 'md4', defaultMessage: 'dd', locale: 'zz-zz' },
        ],
    },
];

describe('lib/riw/app/status/todo', () => {
    it('returns null for missing file', () => {
        mock({
            fixtures: {
                dir: {},
            },
        });

        const received = todo(configResolve(cfgBase));

        mock.restore();

        expect(received).toEqual({ todo: null });
    });

    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                fixtures: {
                    dir: {
                        'TODO.json': stringify(fixture.in),
                    },
                },
            });

            const received = todo(configResolve(cfgBase));

            mock.restore();

            expect(received).toMatchSnapshot();
        });
    });
});
