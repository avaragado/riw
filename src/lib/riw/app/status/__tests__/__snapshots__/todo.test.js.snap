// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/app/status/todo 01 1`] = `
Object {
  "todo": null,
}
`;

exports[`lib/riw/app/status/todo 02 1`] = `
Object {
  "todo": Object {
    "xx-xx": Array [
      Object {
        "defaultMessage": "aa",
        "description": "aa desc",
        "file": "/foo/bar.js",
        "id": "md1",
        "locale": "xx-xx",
      },
      Object {
        "defaultMessage": "bb",
        "description": "bb desc",
        "file": "/foo/bar.js",
        "id": "md2",
        "locale": "xx-xx",
      },
    ],
    "yy-yy": Array [
      Object {
        "defaultMessage": "aa",
        "description": "aa desc",
        "file": "/foo/bar.js",
        "id": "md1",
        "locale": "yy-yy",
      },
    ],
    "zz-zz": Array [
      Object {
        "defaultMessage": "aa",
        "description": "aa desc",
        "file": "/foo/bar.js",
        "id": "md1",
        "locale": "zz-zz",
      },
      Object {
        "defaultMessage": "cc",
        "description": "cc desc",
        "file": "/foo/bar.js",
        "id": "md3",
        "locale": "zz-zz",
      },
      Object {
        "defaultMessage": "dd",
        "file": "/foo/bar.js",
        "id": "md4",
        "locale": "zz-zz",
      },
    ],
  },
}
`;
